walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,120 @@
+* For example, if 4pm - 8pm is free for a meeting, 
+* the Time Range object will represent that entire chunk of time rather than 4pm - 5pm, even if the duration of meeting is 1 hour. 
+*/
 public final class FindMeetingQuery {
+  public static final int MIN_INCREMENT = 10;
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long mtngDuration = request.getDuration();
+    Collection<String> attendees = request.getAttendees();
+    Collection<Event> allEvents = events;
+
+    HashMap<Integer, Integer> mandatoryEvents = makeMandatoryEventMap(allEvents, attendees);
+    Collection<TimeRange> meetingTimes = findMeetingTimes(mtngDuration, mandatoryEvents);
+    return meetingTimes;
+  }
+
+
+  /** 
+  * Create a hash map <event start time, event end time> that ONLY contains meetings where 1 or more people from attendees must attend. 
+  **/
+  public HashMap<Integer, Integer> makeMandatoryEventMap (Collection<Event> allEvents, Collection<String> attendees) {
+    HashMap<Integer, Integer> mandatoryEvents = new HashMap<Integer, Integer>();
+
+    for (Event event : allEvents) {
+      //people who must attend 
+      Set<String> people = event.getAttendees();
+      //check if any people in meeting request is present in this set
+      for (String person : attendees) {
+        if (people.contains(person)) {
+          // If there's already an event with this start time in the map, but the current end time is later, add the current instead.  
+          int eventStartTime = event.getWhen().start();
+          int eventEndTime = event.getWhen().end();
+          if (mandatoryEvents.containsKey(eventStartTime)) {
+            if (eventEndTime > mandatoryEvents.get(eventStartTime)) {
+              mandatoryEvents.put(eventStartTime, eventEndTime);
+            }
+            else {
+              break;
+            }
+          }
+          else {
+            mandatoryEvents.put(eventStartTime, eventEndTime);
+          }
+        }  
+      }
+    }
+
+    return mandatoryEvents;
+  }
+
+  /**
+  * Create and return an ArrayList of TimeRange objects that span available meeting times.
+  */
+  public Collection<TimeRange> findMeetingTimes (long mtngDuration, HashMap<Integer, Integer> mandatoryEvents) {
+    Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+    int currTime = TimeRange.START_OF_DAY;
+    int startTime = TimeRange.START_OF_DAY;
+    int currDuration = 0;
+
+    while (currTime < TimeRange.END_OF_DAY) {
+      // If there's an event starting at this time, compare the current duration and meeting duration to see if there is a possible meeting time. 
+      // Reset the current duration variable, move the current time to the end time of the event, and repeat loop. 
+      if (mandatoryEvents.containsKey(currTime)) {
+        int longestMeetingEnd = mandatoryEvents.get(currTime);
```

Comment:
Does need to be declared here?  It's immediately re-assigned on line 97.

Reply:
I  need to keep track of the current longest meeting end because I'm comparing against it in the isConflict function, but I can call this variable something else to make it clearer